ublock-origin: Init at 1.33.2

chkno opened the pull request:
###### Motivation for this change
Make ublock-origin available.

ungoogled-chromium cannot directly use the Google Chrome Web Store.  Unlike other, simpler Chromium extensions, ublock-origin requires a build step before it can be loaded into Chromium as an unpacked extension.  So this package makes it easier to use ublock-origin with ungoogled-chromium.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

chkno committed:
ublock-origin-chromium: Init at 1.33.2

ublock-origin-firefox: Init at 1.33.2

r-rmcgibbo commented:
Result of `nixpkgs-review pr 113976` at 9e4bf8bd run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/727a6a506565b68d7929a78a6c797ae0">ublock-origin-chromium</a></li></li>
    <li><a href="https://gist.github.com/3084f31a9afaf65f3940758b7b9cad49">ublock-origin-firefox</a></li></li>
  </ul>
</details>

lourkeur commented:
For the Chromium output, could you explain how you would install it into an Ungoogled Chromium profile?

For the Firefox output, it doesn't drop into the `programs.firefox.extensions` option of [home-manager](https://github.com/nix-community/home-manager). It seems like the .xpi needs to be in a specific subdirectory for it to work. You don't necessarily have to support that if that's not your goal though.

lourkeur commented:
The .xpi for Firefox requires fiddling with about:config because it isn't [signed](https://support.mozilla.org/en-US/kb/add-on-signing-in-firefox), but appears to work fine otherwise

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -750,6 +750,9 @@ in
 
   metapixel = callPackage ../tools/graphics/metapixel { };
 
+  inherit (callPackages ../tools/networking/ublock-origin { })
+    ublock-origin-chromium ublock-origin-firefox;
```
```suggestion
  inherit (callPackages ../tools/networking/ublock-origin { }) ublock-origin-chromium ublock-origin-firefox;
```

SuperSandro2000 commented on pkgs/tools/networking/ublock-origin/default.nix:
```diff
@@ -0,0 +1,45 @@
+    sourceRoot = "uBlock";
+    nativeBuildInputs = [ python3 zip ];
+    meta = {
+      description =
+        "uBlock Origin - An efficient blocker for Chromium and Firefox. Fast and lean.";
```
```suggestion
      description = "uBlock Origin - An efficient blocker for Chromium and Firefox. Fast and lean.";
```

SuperSandro2000 commented on pkgs/tools/networking/ublock-origin/default.nix:
```diff
@@ -0,0 +1,45 @@
+    ];
+    sourceRoot = "uBlock";
+    nativeBuildInputs = [ python3 zip ];
+    meta = {
```
```suggestion
    meta = with lib; {
```

SuperSandro2000 commented on pkgs/tools/networking/ublock-origin/default.nix:
```diff
@@ -0,0 +1,45 @@
+    meta = {
+      description =
+        "uBlock Origin - An efficient blocker for Chromium and Firefox. Fast and lean.";
+      license = lib.licenses.gpl3Plus;
+      platforms = lib.platforms.linux;
+      maintainers = with lib.maintainers; [ chkno ];
```
```suggestion
      license = licenses.gpl3Plus;
      platforms = platforms.linux;
      maintainers = with maintainers; [ chkno ];
```

SuperSandro2000 commented on pkgs/tools/networking/ublock-origin/default.nix:
```diff
@@ -0,0 +1,45 @@
+in {
+  ublock-origin-chromium = stdenvNoCC.mkDerivation (common // {
+    pname = "ublock-origin-chromium";
+    buildPhase = "sh ./tools/make-chromium.sh";
```
```suggestion
    buildPhase = ''
      bash tools/make-chromium.sh
    '';
```

SuperSandro2000 commented on pkgs/tools/networking/ublock-origin/default.nix:
```diff
@@ -0,0 +1,45 @@
+  });
+  ublock-origin-firefox = stdenvNoCC.mkDerivation (common // {
+    pname = "ublock-origin-firefox";
+    buildPhase = "sh ./tools/make-firefox.sh all";
```
```suggestion
    buildPhase = ''
      bash tools/make-firefox.sh all
    '';
```

chkno commented:
@lourkeur - I added a longDescription with notes about how to load the extension in both Chromium and Firefox.  Thanks!

@SuperSandro2000 - I applied your suggestions.  Thanks!

 * I'll use --width=999 with nixfmt from now on to save you that trouble in future.
 * I felt bad about `bash foo.sh` being less locally-readable than `sh foo.sh`, so I switched to `patchShebangs` & direct invocation to avoid making assumptions about which shell to use.

SuperSandro2000 commented:
> * I felt bad about `bash foo.sh` being less locally-readable than `sh foo.sh`, so I switched to `patchShebangs` & direct invocation to avoid making assumptions about which shell to use.

sh in nixpkgs is not sh or dash anyway. I think in this case it is bash or ash but I am not fully sure which one.

lourkeur commented:
I followed the instructions and managed to it working in ungoogled-chromium. It'd be nice if it could be somewhat automated, (not a Nix Chromium expert)  but as far as ungoogled is concerned, it's a net benefit. :+1: 

I tried to use the Firefox derivation in `wrapFirefox` but it silently fails too:
```
with pkgs; wrapFirefox firefox-unwrapped {
    nixExtensions = [
      (ublock-origin-firefox // {
        extid = "uBlock0.firefox";
        })
    ];
  }
```

For the Firefox side, I will argue that, in its current state, it should be dropped or not aliased in top-level. If we include a Firefox addon derivation, it should be at least as good as the `fetchFirefoxAddon` based ones found in the [docs](https://github.com/NixOS/nixpkgs/blob/master/doc/builders/packages/firefox.section.md) and in [NUR](https://github.com/nix-community/nur-combined/blob/eaaaf2add511fe45d85052bbcdd1ef8b6bf2c763/repos/rycee/pkgs/firefox-addons/generated-firefox-addons.nix#L1064). These can be declaratively installed and updated through Nix and/or Home-manager. This derivation does not support that, and afaict has nothing over those. It would be detrimental if the official-looking way to install ublock was worse in every way.

bb2020 commented:
For Chromium, another option is to install with a group policy.

```
environment.etc."chromium/policies/managed/policy.json".text = builtins.toJSON
{
  "ExtensionInstallForcelist" = [ "cjpalhdlnbpafiamejdnhcphjbkeiagm" ];
};
```

lourkeur commented:
> For Chromium, another option is to install with a group policy.
How about ungoogled-chromium? Try
```
nix shell -p ungoogled-chromium
chromium -p $(mktemp -d)
```
Ungoogled chromium intentionally disables the webstore, hence it cant turn "cjpalhdlnbpafiamejdnhcphjbkeiagm" into a usable extension. Unless you run normal chromium first, it wont install or update extensions.

bb2020 commented:
Just wanted to leave a remark for regular Chromium users.

michaeladler commented:
Having to enable developer mode works but is quite inconvenient. I veto against this approach. If the upstream project provides a (signed) crx3 file, it would work out of the box without having to toggle any settings in the browser. browserpass does this and I'm maintaining it in my overlay, see https://github.com/michaeladler/nur-packages/tree/master/pkgs/chromium-browserpass. Extensions like that could be added to nixpkgs I guess. Arch Linux [does it too](https://github.com/archlinux/svntogit-community/blob/packages/browserpass-chromium/trunk/PKGBUILD).

SuperSandro2000 commented:
Closing because being stale and the requiring to activate developer mode is not ideal.
